### PR TITLE
fix: dataset modal doesn't work in dev mode

### PR DESCRIPTION
**Pull request by stephenLYZ**
### SUMMARY

This problem only exists in dev environments. The reason is the react-hot-loader will use "ProxyFacade", which is a wrapper for Functional Component. 

```
ProxyFacade = function(props, context) {
      /*

        ! THIS IS NOT YOUR COMPONENT !
        !  THIS IS REACT-HOT-LOADER  !

        And you are probably looking for a function component of yours
        It's hidden, but there is a way to fix this - just reconfigure your application a bit
        see https://github.com/gaearon/react-hot-loader/issues/1311

       */

      const result = CurrentComponent(props, context);
```

And we use`recurseReactClone` method to create component(see https://github.com/apache/superset/blob/master/superset-frontend/src/CRUD/utils.js), which will pass props to the component. In this case, since the name has been changed, it is not possible to pass `onChange` props. see screenshot:

image

Here we use this solution. https://github.com/gaearon/react-hot-loader/issues/1311

### befre
https://user-images.githubusercontent.com/33317356/153366245-f733f6b6-c7c1-4322-b16a-91eb1d156b1c.mp4
### after

https://user-images.githubusercontent.com/11830681/153376126-886fd66b-9bc1-4d32-bae8-86d1b256e04b.mov

### ADDITIONAL INFORMATION

- [x] Has associated issue: https://github.com/apache/superset/issues/18655
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: superset-frontend/webpack.config.js [hunk 1]
```diff
@@ -344,6 +344,11 @@ const config = {
         ],
         use: [babelLoader],
       },
+      {
```
Comment (zhaoyongjie) [inline superset-frontend/webpack.config.js:347]: Please add a `todo` comment here and link to this [PR](https://github.com/apache/superset/issues/18655). We should refactor `DatasourceEditor.jsx` and decouple it in future.

File: superset-frontend/webpack.config.js [hunk 2]
```diff
@@ -344,6 +344,11 @@ const config = {
...
+        test: /\.js$/,
+        include: /node_modules\/react-dom/,
+        use: ['react-hot-loader/webpack'],
+      },
```
[the remark this answers is not in the document]
Comment (stephenLYZ) [reply superset-frontend/webpack.config.js:351]: Good idea. However, this [lib](https://github.com/hot-loader/react-dom) has not been updated for three years. 😂

**Review by zhaoyongjie — APPROVED** [review]
LGTM

File: superset-frontend/webpack.config.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (villebro) [reply superset-frontend/webpack.config.js:351]: Uhh, good point! 😓

**Review by villebro — APPROVED** [review]
LGTM, thanks for fixing this!

File: superset-frontend/webpack.config.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (stephenLYZ) [reply superset-frontend/webpack.config.js:351]: FYI. We may use [react-refresh](https://github.com/pmmmwh/react-refresh-webpack-plugin) to replace react-hot-loader in the future, which is more active.

**Pull request merged**